### PR TITLE
[Feat]support horovod sync train

### PR DESCRIPTION
**Pull request by a6802739**
- Use horovod to sync dense grad of the model parameter,  for `TrainableWrapper`'s grad, keep same with before.

### CODE REVIEW

**Comment by rhdong** [comment]
Hi @a6802739 , thank you for your contribution!

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# A distributed training demo for `tfra.dynamic_embedding`:
```
Comment (rhdong) [inline demo/dynamic_embedding/movielens-100k-sync-estimator/README.md:1]: Changing name to movielens-100k-sync-estimator-with-horovod would be better.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
 import functools
 from tensorflow_recommenders_addons.dynamic_embedding.python.ops.dynamic_embedding_creator import KVCreator
 import six
+import horovod.tensorflow as hvd
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:24]: Please format all the files firstly, refer to: https://github.com/tensorflow/recommenders-addons/blob/master/STYLE_GUIDE.md

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 3]
```diff
@@ -273,11 +277,65 @@ def _zeros_slot(var, slot_name, op_name):
...
+      return train_op
+
   if isinstance(self, optimizer.Optimizer):
     self._get_or_make_slot = _get_or_make_slot
     self._get_or_make_slot_with_initializer = _get_or_make_slot_with_initializer
     self._zeros_slot = _zeros_slot
+    self.apply_gradients = apply_gradients
   elif isinstance(self, optimizer_v2.OptimizerV2):
+    if self._dist_sync:
+      raise Exception("OptimizerV2 didn't support distributed sync train now, please use Optimizer")
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:338]: please use 'tf.train.XxxxOptimizer'

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 4]
```diff
@@ -39,7 +40,7 @@
 from tensorflow.python.training import slot_creator
 
 
-def DynamicEmbeddingOptimizer(self, bp_v2=None):
+def DynamicEmbeddingOptimizer(self, bp_v2=None, dist_sync=None):
   """ An optimizer wrapper to make any TensorFlow optimizer capable of training
   Dynamic Embeddding Variables.
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:46]: Please add comments for the two parameters, I give you the bp_v2's comment:
"""
     Args:
      bp_v2: If True, updating parameters will use *updating* instead of *setting*, which solves
        the race condition problem among workers during back-propagation in large-scale
        distributed asynchronous training.  
        Reference: [Jiang et al., 2020](https://www.usenix.org/conference/osdi20/presentation/jiang)
      dist_sync: bulabula....
"""

Comment (rhdong) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:46]: Please pay attention to the indentation

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 5]
```diff
@@ -39,7 +40,7 @@
 from tensorflow.python.training import slot_creator
 
 
-def DynamicEmbeddingOptimizer(self, bp_v2=None):
+def DynamicEmbeddingOptimizer(self, bp_v2=None, dist_sync=None):
```
Comment (rhdong) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:43]: ‘dist_sync’ is a little confusing words. Maybe 'synchronous' or 'use_allreduce' would be better?
BTW, the default should be False?

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+- dataset: [movielen/100k-ratings](https://www.tensorflow.org/datasets/catalog/movielens#movielens100k-ratings)
+- model: DNN
+- Running API: using estimator APIs
+
```
Comment (rhdong) [inline demo/dynamic_embedding/movielens-100k-sync-estimator/README.md:6]: Please add version requirements for Horvod and MPI.

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md (hunk 1 above)
Comment (rhdong) [inline demo/dynamic_embedding/movielens-100k-sync-estimator/README.md:1]: A distributed synchronous training demo based on Horovod.

File: demo/dynamic_embedding/movielens-100k-sync-estimator/export.sh [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+rm -rf ./export_dir
+sh stop.sh
+
+sleep 1
+export TF_CONFIG='{"cluster": {"worker": ["localhost:2222"], "ps": ["localhost:2223"], "chief": ["localhost:2224"]}, "task": {"type": "chief", "index": 0}}'
```
Comment (rhdong) [inline demo/dynamic_embedding/movielens-100k-sync-estimator/export.sh:6]: Maybe you should modify this file according to you training pipeline?

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md (hunk 1 above)
Comment (a6802739) [reply demo/dynamic_embedding/movielens-100k-sync-estimator/README.md]: done.

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md (hunk 6 above)
Comment (a6802739) [reply demo/dynamic_embedding/movielens-100k-sync-estimator/README.md]: done.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 5 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: change to `horovod_synchronous` here.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 4 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:46]: Thanks, done.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 3 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: done.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 2 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: done.

File: demo/dynamic_embedding/movielens-100k-sync-estimator/README.md (hunk 1 above)
Comment (a6802739) [reply demo/dynamic_embedding/movielens-100k-sync-estimator/README.md]: done.

File: demo/dynamic_embedding/movielens-100k-sync-estimator/export.sh (hunk 7 above)
Comment (a6802739) [reply demo/dynamic_embedding/movielens-100k-sync-estimator/export.sh]: done.

File: demo/dynamic_embedding/movielens-100k-sync-estimator-with-horovod/session_hook.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import horovod.tensorflow as hvd
```
Comment (Lifann) [inline demo/dynamic_embedding/movielens-100k-sync-estimator-with-horovod/session_hook.py:1]: It will introduce a new hvd dependency. Maybe use [NcclAllReduce](https://www.tensorflow.org/api_docs/python/tf/distribute/NcclAllReduce#broadcast) directly could avoid additional dependencies, isolated communicator problem, and cost of negotiation in horovod.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 9]
```diff
@@ -19,6 +19,7 @@
 from __future__ import print_function
 
 import functools
+import horovod.tensorflow as hvd
```
Comment (Lifann) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:22]: It will introduce a new hvd dependency. Maybe use NcclAllReduce directly could avoid additional dependencies, isolated communicator problem, and performance cost of negotiation in horovod.

Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:22]: I think it will be better to use horovod here. For CPU device, horovod will user `MPI_AllReduce`, for GPU device, horovod will use `NcclAllReduce`.

Comment (Lifann) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:22]: Right. Some may do standalone or asynchronous training, without horovod. I think it would be better to only import horovod while using the synchronous training. Thus make the horovod  unbound.

Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:22]: Hmm, I see. Thanks for your suggestion here.

Comment (Lifann) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:22]: Maybe the following code can help.
```python
try:
  import horovod.tensorflow as hvd
except ImportError:
  pass

hvd.allreduce(...) # It will raise module not found error if hvd not installed.
```

Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: Thanks. Done now.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 10]
```diff
@@ -41,14 +41,17 @@
 from tensorflow.python.training.tracking import base as trackable
 
 
-def DynamicEmbeddingOptimizer(self, bp_v2=None):
+def DynamicEmbeddingOptimizer(self, bp_v2=False, horovod_synchronous=False):
```
Comment (Lifann) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:44]: The name of `horovod_synchronous` may bind the API to horovod only. I think `synchronous` might be better.

File: tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/horovod_sync_train_test.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+      test_loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=y,
+                                                          logits=test_logits)
+
+      grads_and_vars = test_opt.compute_gradients(test_loss,
+                                                  var_list=[test_weight])
+      var_list = []
+      aggregated_grad = []
+      for grad, var in grads_and_vars:
+        var_list.append(var)
+        aggregated_grad.append(hvd.allreduce(grad, op=hvd.Sum))
```
Comment (Lifann) [inline tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/horovod_sync_train_test.py:85]: And just for discussion, I noticed that Horovod implements lots of features and optimizations on its DistributedOptimizer like tensor fusion, grouped allreduce, adasum, gradient compression, etc.., in the latest-nth versions. Use the hvd.allreduce API may not reuse these features.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 12]
```diff
@@ -281,12 +287,77 @@ def _zeros_slot(var, slot_name, op_name):
...
+          trainable_grad_and_vars.append((grad, var))
+          continue
+        var_list.append(var)
+        with ops.device(var.device):
+          # Dense gradients.
+          if grad is None:
+            aggregated_grad.append(None)  # pass-through.
+            continue
+          elif isinstance(grad, ops.Tensor):
+            aggregated_grad.append(hvd.allreduce(grad, op=hvd.Sum))
```
Comment (Lifann) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:336]: And just for discussion, I noticed that Horovod implements lots of features and optimizations on its DistributedOptimizer like tensor fusion, grouped allreduce, adasum, gradient compression, etc.., in the latest-nth versions. Use the hvd.allreduce API may not reuse these features.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py [hunk 13]
```diff
@@ -281,12 +287,77 @@ def _zeros_slot(var, slot_name, op_name):
...
+        if isinstance(var, de.TrainableWrapper):
+          trainable_grad_and_vars.append((grad, var))
+          continue
+        var_list.append(var)
+        with ops.device(var.device):
+          # Dense gradients.
+          if grad is None:
+            aggregated_grad.append(None)  # pass-through.
+            continue
+          elif isinstance(grad, ops.Tensor):
```
Comment (Lifann) [inline tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:335]: MLP layers, downstream from embedding, may also generate IndexedSlices gradient, need to notice that.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 5 above)
Comment (Lifann) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: The name of `horovod_synchronous` may bind the API to horovod only. `synchronous` might be enough?

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 12 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:337]: Yeah, For `tensor fusion`, it's set by environment variable, and it's turn on by default, for most recommend situation, I think there is no need to care about `grouped allreduce`  or `gradient compression`. But I think we could find better way to open `reduction operation` to the user. If we directly let user specify `hvd.Sum` or `hvd.adamsum`, we should let them import `horovod` before `apply_gradients`, or we could let user specify a string the reduction method like `sum`, we map this method to the corresponding `hvd op`, but if  horovod add some reduction op, we should change the code to be compatible with newest horovod version.

File: tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py (hunk 13 above)
Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:336]: Sorry, I never notice MLP layers  could generate IndexedSlices gradient, could you give me an example?

Comment (Lifann) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:336]:
```

Comment (Lifann) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py:336]: > Sorry, I never notice MLP layers could generate IndexedSlices gradient, could you give me an example?

For example:

```python
var = tf.Variable(...)
params = de.get_variable(...)
emb = de.embedding_lookup(...)
latent_tensor = sum_pooling(emb)
... = some_func(latent_tensor, var)
```

The `some_layer` are defined as:
```python
def some_func(latent, var):
  mask = tf.greater_equal(var, threshold)
  pos = tf.where(mask)
  selected = tf.gather(var, pos)
  return tf.reduce_mean(selected)
```

The `some_func` code will make the gradient become IndexedSlices to `var`.

Comment (a6802739) [reply tensorflow_recommenders_addons/dynamic_embedding/python/ops/dynamic_embedding_optimizer.py]: I see, Thanks. Now I delelte the judgement `isinstance(grad, ops.Tensor)`. So Horovod will check it's grad type and deal with it.

**Review by Lifann — APPROVED** [review]
LGTM
